hutch_python/epics_arch.py

```diff
@@ -93,13 +116,138 @@ def create_arch_file(experiment, hutch=None, path=None, dry_run=False):
+    if len(af_data) > 0:
+        # ArchFile Data, removing whitespaces and newline chars
+        afDict = dict(zip(af_data[::2], af_data[1::2]))
+        afDict = {k.replace(" ", ""): v for k, v in afDict.items()}
+        afDict = {k.replace("\n", ""): v for k, v in afDict.items()}
+        afDict = {k: v.replace(" ", "") for k, v in afDict.items()}
+        afDict = {k: v.replace("\n", "") for k, v in afDict.items()}
+        sorted_afDict = dict(sorted(afDict.items()))
+
+    # PART 2
+
+    # Check the questionaire for duplicate PVs
+    # Making reverse multidict to help identify duplicate values in questionnaire.
+    rev_keyDict = {}
+    for key, value in sorted_qsDict.items():
+        rev_keyDict.setdefault(value, list()).append(key)
+
+    pvDuplicate = [key for key, values in rev_keyDict.items() if len(values) > 1]
+    # Looking for duplicates of PVs in the questionaire
+    # also print out the alias for PV, change removing to warning operater to remove dup then rerun
+    for dup in pvDuplicate:
+        logger.debug("!Duplicate PV in questionnaire!:" + str(dup))
+        for value in rev_keyDict[dup][1:]:
+            logger.debug("Found PV duplicate(s) from questionnaire: " + value + ", " + sorted_qsDict[value])
+        raise Exception("Please remove duplicates and re-run script!")
+
+    # Check to see if the archfile has any data in it
+    if len(af_data) == 0:
+        logger.debug("CFD: Case: no archfile given, returning cleaned questionnaire data\n")
+        cleaned_qs_data = [x for item in sorted_qsDict.items() for x in item]
+        return cleaned_qs_data
+
+    # Once we have cleared any duplicates in the questionnaire we moving on to updating values according to the which field matches.
+
+    # Checking for matching PVs in questionnaire and archfile
+    # if the PV matches update the alias by removing the old key and making a new one
+    for (k, val) in sorted_qsDict.items():
+        # this looks up the key in the af Dictionary by finding the value
+        foundKey = get_key(val, sorted_afDict)
+        if k in sorted_afDict:
+            logger.debug("!Alias Match in questionnaire and archfile! Updating PV: " + k + ", " + sorted_qsDict[k])
+            sorted_afDict[k] = sorted_qsDict[k]
+        elif foundKey:
+            del sorted_afDict[foundKey]
+            sorted_afDict[k] = val
+            logger.debug("!PV Match in questionnaire and archfile! Updating Alias: " + k + ", " + val)
+
+    sorted_afDict = dict(sorted(sorted_afDict.items()))
+    updated_arch_list = [x for item in sorted_afDict.items() for x in item]
+    logger.debug("\nUpdated Arch List:\n")
+    logger.debug(updated_arch_list)
+    return updated_arch_list
+
+
+def read_archfile(exp_path):
+    if os.path.exists(exp_path):
+        with open(exp_path, "r") as experiment:
+            lines = experiment.readlines()
+        return lines
+    else:
+        raise OSError('ArchFile not found: %s' % exp_path)
```

Comment:
Q: What are we guarding against by raising our own OSError instead of using the error raised by open?